text/1011-query-rewrite.rst

```diff
@@ -0,0 +1,322 @@
+      ALTER ANNOTATION <annotation-name> := <value>
+      DROP ANNOTATION <annotation-name>
+
+
+DROP POLICY
+-----------
+
+Remove a query rewrite policy.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    ALTER OBJECT TYPE <type-name> "{"
+        DROP POLICY
+            ON { SELECT | INSERT | UPDATE | DELETE } [, ... ]
+            [ WHEN <condition> ]
+            <action>
+    "}"
+
+
+Interaction with globals
+========================
+
+Query rewrite policies are especially powerful when combined with RFC 1010
+globals, because then data visibility can be globally adjusted with a single
+``SET GLOBAL`` statement, which is very useful for authenticated/authorized
+data access control.
+
+Example::
+
+    global user_id -> uuid;
+
+    abstract object type Owned {
+      required link owner -> User;
+
+      policy on select, update, delete
+        permit (.owner.id = global user_id)
+
+      policy on insert, update
+        check (.owner.id = global user_id)
+    }
+
+    object type Purchase extending Owned;
+
+    ...
+
+    set global user_id := <uuid-1>;
+    select count(Purchase);
+    # 9
+    set global user_id := <uuid-2>
+    select count(Purchase);
+    # 1
+
+
+Another example::
+
+    abstract object type LoggedAccess {
+      policy on select
+        do after (for obj in __result__ union (
+          insert AccessLog {
+            object := obj,
+            user_id := global user_id,
+            time := datetime_current(),
+          }
+        ))
+    }
+
+
+Bypassing policies
+==================
+
+A superuser can bypass the execution of query rewrite policies by setting
+the ``apply_rewrite_policies`` session configuration setting to ``false``.
+
+
+Introspection
+=============
+
+Policies can be introspected via a new ``schema::RewritePolicy`` in the
+introspection schema that is linked from ``schema::ObjectType`` via the new
+``policies`` link.  The ``schema::RewritePolicy`` is exposed as follows::
+
+    type schema::RewritePolicy extending schema::AnnotationSubject {
+      multi property statements -> schema::StatementKind;
+      property condition -> std::str;
+      required property action_kind -> schema::PolicyActionKind;
+      required property action -> std::str;
+    };
+
+
+Implementation considerations
+=============================
+
+Like the title of the RFC suggests, policies are a query rewrite mechanism,
+which means that they primarily affect what IR is generated for a given
+EdgeQL query.  All actions except ``DO ON COMMIT`` are pure EdgeQL transforms.
+
+``PERMIT`` and ``RESTRICT TO`` wrap set references and transform every ``Foo``
+reference into ``(SELECT Foo FILTER <permit-restrict-filter>)``.
+
+``CHECK`` actions insert an intermediate shape into ``INSERT`` and ``UPDATE``,
+e.g.::
+
+    INSERT Foo { prop := <value> }
+
+is roughly transformed into::
+
+    WITH
+      input := { prop := <value> },
+      checked := input { prop := <check_expr> }
```

Comment:
how does it make the query error out when check_expr doesn't evaluate to any `true`?

Reply:
Yeah, sorry, it's actually expanded to something like

```
  checked := input { prop := prop IF (SELECT _ := <check_expr> FILTER _) ELSE raise() }
```